File: src/actions/rotary.ts
/*
 * xp_streamdeck - Stream Deck plugin for X-Plane 12
 * Copyright (c) 2026 thWelly
 *
 * Licensed under the MIT License.
 * See the LICENSE file in the project root for full license text.
 */

import streamDeck, {
	action,
	type DidReceiveSettingsEvent,
	type KeyDownEvent,
	type KeyUpEvent,
	SingletonAction,
	type WillAppearEvent,
	type WillDisappearEvent,
} from "@elgato/streamdeck";
import type { JsonObject } from "@elgato/utils";

import { selectors } from "../selectors/registry";
import { coerceNumber, toFiniteNumber } from "../util/coerce";
import { applyIndex, parseDataRefPath } from "../util/dataref-path";
import { parseEnumMap } from "../util/enum";
import { clearOffline, combineTitle, NOT_FOUND_SUFFIX, setOffline } from "../util/error-tile";
import { formatDataRefValue } from "../util/format";
import { extractPlaceholderKeys, substitutePlaceholders } from "../util/placeholders";
import { normalizeFormat, resolveZeroSnap, trimString } from "../util/settings";
import { applyStep } from "../util/step";
import type { DataRefValue, SubscriptionHandle, XPlaneClient } from "../xplane";

type RotaryDirection = "left" | "right" | "up" | "down";
type RotaryFormatMode = "numeric" | "enum";

type RotarySettings = JsonObject & {
	commandPath?: string;
	hideConfirmation?: boolean;
	hideEndstopAlert?: boolean;
	direction?: RotaryDirection;
	datarefPath?: string;
	delta?: string | number;
	minValue?: string | number;
	maxValue?: string | number;
	cycle?: boolean;
	label?: string;
	formatMode?: RotaryFormatMode;
	format?: string;
	unit?: string;
	unitScale?: string | number;
	precision?: string | number;
	enumMap?: string;
	holdOnLastPosition?: boolean;
	holdCommand?: string;
	snapZero?: boolean;
	zeroThreshold?: string | number;
};

interface ParsedSettings {
	commandPath: string;
	datarefPath: string;
	delta?: number;
	sign: 1 | -1;
	minValue?: number;
	maxValue?: number;
	cycle: boolean;
	hideEndstopAlert: boolean;
	label: string;
	formatMode: RotaryFormatMode;
	format: string;
	unit: string;
	unitScale?: number;
	precision?: number;
	enumLut: Map<number, string>;
	enumMaxIndex: number | undefined;
	enumValid: boolean;
	holdOnLastPosition: boolean;
	holdCommand: string;
	zeroSnap?: number;
}

interface ActionState {
	action: WillAppearEvent<RotarySettings>["action"];
	parsed: ParsedSettings;
	handle?: SubscriptionHandle;
	lastValue?: DataRefValue;
	holdInProgress: boolean;
	activeHoldId?: number;
}

@action({ UUID: "com.robertw.xplane.rotary" })
export class XPlaneRotary extends SingletonAction<RotarySettings> {
	private readonly states = new Map<string, ActionState>();

	constructor(private readonly xplane: XPlaneClient) {
		super();
		this.xplane.on("offline", () => this.onXPlaneOffline());
		this.xplane.on("online", () => this.onXPlaneOnline());
		selectors.watch((changed) => this.onSelectorsChanged(changed));
	}

	override async onWillAppear(ev: WillAppearEvent<RotarySettings>): Promise<void> {
		const parsed = parseSettings(ev.payload.settings ?? {});
		const state: ActionState = {
			action: ev.action,
			parsed,
			holdInProgress: false,
		};
		this.states.set(ev.action.id, state);

		if (this.xplane.isOffline()) {
			await setOffline(ev.action);
			return;
		}
		await this.applySubscription(state);
	}

	override onWillDisappear(ev: WillDisappearEvent<RotarySettings>): Promise<void> {
		const state = this.states.get(ev.action.id);
		if (!state) return Promise.resolve();
		this.dropSubscription(state);
		// If a hold was still in progress (unusual — usually onKeyUp clears it),
		// best-effort release so X-Plane doesn't get stuck with the starter held.
		if (state.holdInProgress && state.activeHoldId !== undefined) {
			const id = state.activeHoldId;
			this.xplane
				.endCommand(id)
				.catch((err) =>
					streamDeck.logger.warn("rotary: endCommand on disappear failed", err),
				);
			state.holdInProgress = false;
			state.activeHoldId = undefined;
		}
		this.states.delete(ev.action.id);
		return Promise.resolve();
	}

	override async onDidReceiveSettings(
		ev: DidReceiveSettingsEvent<RotarySettings>,
	): Promise<void> {
		const state = this.states.get(ev.action.id);
		if (!state) return;

		const next = parseSettings(ev.payload.settings ?? {});
		const pathChanged = next.datarefPath !== state.parsed.datarefPath;
		state.parsed = next;

		if (pathChanged) {
			this.dropSubscription(state);
			state.lastValue = undefined;
			if (!state.parsed.datarefPath) {
				await state.action.setTitle(state.parsed.label);
			}
			await this.applySubscription(state);
			return;
		}

		this.render(state);
	}

	override async onKeyDown(ev: KeyDownEvent<RotarySettings>): Promise<void> {
		const state = this.states.get(ev.action.id);
		const parsed = parseSettings(ev.payload.settings ?? {});
		const hideConfirmation = ev.payload.settings?.hideConfirmation === true;

		const canStep = parsed.delta !== undefined && !!parsed.datarefPath;
		if (!parsed.commandPath && !parsed.holdCommand && !canStep) {
			streamDeck.logger.warn(
				"rotary: commandPath / holdCommand / delta+datarefPath are all empty",
			);
			await ev.action.showAlert();
			return;
		}

		const snap = selectors.snapshot();

		if (shouldHoldOnLast(state, parsed)) {
			const holdCommand = substitutePlaceholders(parsed.holdCommand, snap);
			try {
				const id = await this.xplane.getCommandId(holdCommand);
				await this.xplane.beginCommand(id);
				if (state) {
					state.holdInProgress = true;
					state.activeHoldId = id;
				}
				streamDeck.logger.info(`rotary: hold begin ${holdCommand} (id=${id})`);
			} catch (err) {
				streamDeck.logger.error(`rotary: hold begin failed: ${holdCommand}`, err);
				await ev.action.showAlert();
			}
			return;
		}

		// Optional DataRef step (cycle/clamp) — takes precedence over command activate.
		if (canStep && state) {
			await this.stepDataRef(state, parsed, hideConfirmation);
			return;
		}

		if (parsed.formatMode === "enum" && !parsed.enumValid) {
			streamDeck.logger.warn("rotary: enumMap parse error — refusing to fire command");
			await ev.action.showAlert();
			return;
		}

		if (!parsed.commandPath) {
			streamDeck.logger.warn("rotary: commandPath empty (HOLD / step did not apply)");
			await ev.action.showAlert();
			return;
		}

		const commandPath = substitutePlaceholders(parsed.commandPath, snap);
		try {
			const id = await this.xplane.getCommandId(commandPath);
			await this.xplane.activateCommand(id);
			streamDeck.logger.info(`rotary: activate ${commandPath} (id=${id})`);
			if (!hideConfirmation) {
				await ev.action.showOk();
			}
		} catch (err) {
			streamDeck.logger.error(`rotary: command failed: ${commandPath}`, err);
			await ev.action.showAlert();
		}
	}

	private async stepDataRef(
		state: ActionState,
		parsed: ParsedSettings,
		hideConfirmation: boolean,
	): Promise<void> {
		const delta = parsed.delta;
		if (delta === undefined || !(delta > 0) || !parsed.datarefPath) {
			await state.action.showAlert();
			return;
		}
		try {
			const resolvedPath = substitutePlaceholders(parsed.datarefPath, selectors.snapshot());
			const { basePath, index } = parseDataRefPath(resolvedPath);
			const drId = await this.xplane.getDataRefId(basePath);
			const currentRaw =
				state.lastValue !== undefined
					? state.lastValue
					: applyIndex(await this.xplane.readDataRef(drId), index);
			const current = coerceNumber(currentRaw) ?? 0;
			const { value: target, blocked } = applyStep(current, parsed.sign * delta, {
				min: parsed.minValue,
				max: parsed.maxValue,
				cycle: parsed.cycle,
			});
			if (blocked) {
				streamDeck.logger.info(
					`rotary: step at endstop for ${resolvedPath} (value=${current})`,
				);
				if (!parsed.hideEndstopAlert) await state.action.showAlert();
				return;
			}
			await this.xplane.writeDataRef(drId, target, index);
			streamDeck.logger.info(`rotary: step ${resolvedPath} ${current} → ${target}`);
			state.lastValue = target;
			this.render(state);
			if (!hideConfirmation && state.action.isKey()) await state.action.showOk();
		} catch (err) {
			streamDeck.logger.error("rotary: dataref step failed", err);
			await state.action.showAlert();
		}
	}

	override async onKeyUp(ev: KeyUpEvent<RotarySettings>): Promise<void> {
		const state = this.states.get(ev.action.id);
		if (!state?.holdInProgress || state.activeHoldId === undefined) return;
		const id = state.activeHoldId;
		const hideConfirmation = ev.payload.settings?.hideConfirmation === true;
		try {
			await this.xplane.endCommand(id);
			streamDeck.logger.info(`rotary: hold end (id=${id})`);
			if (!hideConfirmation) {
				await ev.action.showOk();
			}
		} catch (err) {
			streamDeck.logger.error(`rotary: endCommand failed (id=${id})`, err);
			await ev.action.showAlert();
		} finally {
			state.holdInProgress = false;
			state.activeHoldId = undefined;
		}
	}

	private async applySubscription(state: ActionState): Promise<void> {
		if (!state.parsed.datarefPath) {
			await state.action.setTitle(state.parsed.label);
			return;
		}
		if (this.xplane.isOffline()) {
			await setOffline(state.action);
			return;
		}
		const resolved = substitutePlaceholders(state.parsed.datarefPath, selectors.snapshot());
		const { basePath, index } = parseDataRefPath(resolved);
		try {
			state.handle = await this.xplane.subscribe(basePath, (raw) => {
				try {
					state.lastValue = applyIndex(raw, index);
					this.render(state);
				} catch (err) {
					streamDeck.logger.warn(
						`rotary: index apply failed for ${state.parsed.datarefPath}`,
						err,
					);
					state.action
						.setTitle(combineTitle(state.parsed.label, NOT_FOUND_SUFFIX))
						.catch((e) => streamDeck.logger.warn("rotary: setTitle failed", e));
				}
			});
		} catch (err) {
			streamDeck.logger.warn(`rotary: subscribe failed for ${state.parsed.datarefPath}`, err);
			await state.action.setTitle(combineTitle(state.parsed.label, NOT_FOUND_SUFFIX));
			await state.action.showAlert();
		}
	}

	private dropSubscription(state: ActionState): void {
		if (state.handle) {
			this.xplane.unsubscribe(state.handle);
			state.handle = undefined;
		}
	}

	private render(state: ActionState): void {
		if (state.lastValue === undefined) return;
		const { parsed } = state;
		let valueText: string;
		if (parsed.formatMode === "enum") {
			if (!parsed.enumValid) {
				valueText = NOT_FOUND_SUFFIX;
			} else {
				const idx = toIndexInteger(state.lastValue, parsed.unitScale);
				if (idx === undefined) {
					valueText = NOT_FOUND_SUFFIX;
				} else {
					const label = parsed.enumLut.get(idx);
					if (label === undefined) {
						// Index falls between defined detents (e.g. flaps still
						// travelling between 10° and 20°). Skip the title update
						// so the last known detent stays visible until the new
						// one settles — better than flashing "?(N)" through the
						// transition.
						return;
					}
					valueText = label;
				}
			}
		} else {
			const formatted = formatDataRefValue(state.lastValue, {
				format: parsed.format,
				unitScale: parsed.unitScale,
				precision: parsed.precision,
				zeroSnap: parsed.zeroSnap,
			});
			valueText = parsed.unit ? `${formatted} ${parsed.unit}` : formatted;
		}
		state.action
			.setTitle(combineTitle(parsed.label, valueText))
			.catch((err) => streamDeck.logger.warn("rotary: setTitle failed", err));
	}

	private onXPlaneOffline(): void {
		for (const state of this.states.values()) {
			// X-Plane went away mid-hold: best-effort release. WS is gone, but a
			// graceful close may still answer briefly. Clear the local flag
			// either way so a later keyUp doesn't try to end again.
			if (state.holdInProgress && state.activeHoldId !== undefined) {
				this.xplane.endCommand(state.activeHoldId).catch(() => {
					/* sim offline — nothing else to do */
				});
				state.holdInProgress = false;
				state.activeHoldId = undefined;
			}
			this.dropSubscription(state);
			state.lastValue = undefined;
			setOffline(state.action).catch((err) =>
				streamDeck.logger.warn("rotary: setOffline failed", err),
			);
		}
	}

	private onXPlaneOnline(): void {
		for (const state of this.states.values()) {
			clearOffline(state.action)
				.then(() => this.applySubscription(state))
				.catch((err) => streamDeck.logger.warn("rotary: re-subscribe failed", err));
		}
	}

	private onSelectorsChanged(changed: ReadonlySet<string>): void {
		for (const state of this.states.values()) {
			if (!state.parsed.datarefPath) continue;
			const keys = extractPlaceholderKeys(state.parsed.datarefPath);
			if (!keys.some((k) => changed.has(k))) continue;
			this.dropSubscription(state);
			state.lastValue = undefined;
			this.applySubscription(state).catch((err) =>
				streamDeck.logger.warn(
					`rotary: selector re-subscribe failed for ${state.parsed.datarefPath}`,
					err,
				),
			);
		}
	}
}

function shouldHoldOnLast(state: ActionState | undefined, parsed: ParsedSettings): boolean {
	if (parsed.formatMode !== "enum") return false;
	if (!parsed.holdOnLastPosition) return false;
	if (!parsed.holdCommand) return false;
	if (parsed.enumLut.size < 2) return false;
	if (parsed.enumMaxIndex === undefined) return false;
	if (!state || state.lastValue === undefined) return false;
	const idx = toIndexInteger(state.lastValue, parsed.unitScale);
	if (idx === undefined) return false;
	// Trigger on the second-to-last index: pressing now would advance to the
	// last position (which is the "hold" position, e.g. starter on Cessna).
	return idx === parsed.enumMaxIndex - 1;
}

function parseSettings(s: RotarySettings): ParsedSettings {
	const formatMode: RotaryFormatMode = s.formatMode === "enum" ? "enum" : "numeric";
	const { enumLut, enumMaxIndex, enumValid } = parseEnumMap(s.enumMap ?? "");
	const deltaRaw = toFiniteNumber(s.delta);
	const delta = deltaRaw !== undefined && deltaRaw > 0 ? deltaRaw : undefined;
	const sign: 1 | -1 = s.direction === "left" || s.direction === "down" ? -1 : 1;
	return {
		commandPath: trimString(s.commandPath),
		datarefPath: trimString(s.datarefPath),
		delta,
		sign,
		minValue: toFiniteNumber(s.minValue),
		maxValue: toFiniteNumber(s.maxValue),
		cycle: s.cycle === true,
		hideEndstopAlert: s.hideEndstopAlert === true,
		label: trimString(s.label),
		formatMode,
		format: normalizeFormat(s.format),
		unit: trimString(s.unit),
		unitScale: toFiniteNumber(s.unitScale),
		precision: toFiniteNumber(s.precision),
		enumLut,
		enumMaxIndex,
		enumValid,
		holdOnLastPosition: s.holdOnLastPosition === true,
		holdCommand: trimString(s.holdCommand),
		zeroSnap: resolveZeroSnap(s.snapZero, s.zeroThreshold),
	};
}

function toIndexInteger(v: DataRefValue, scale?: number): number | undefined {
	const raw = coerceNumber(v);
	if (raw === undefined) return undefined;
	const scaled = scale !== undefined && Number.isFinite(scale) ? raw * scale : raw;
	return Math.round(scaled);
}

File: streamcontroller/com_robertw_xplane/actions/Rotary/Rotary.py
"""
X-Plane Rotary — key/dial step for multi-position switches.

Press either:
  - steps a DataRef by ±Delta (Direction sets the sign), with optional Min/Max/Cycle, or
  - activates a CommandPath when Delta is unset / empty.
"""

from __future__ import annotations

from GtkHelper.GenerativeUI.ComboRow import ComboRow
from GtkHelper.GenerativeUI.EntryRow import EntryRow
from GtkHelper.GenerativeUI.SpinRow import SpinRow
from GtkHelper.GenerativeUI.SwitchRow import SwitchRow
from loguru import logger as log
from src.backend.DeckManagement.InputIdentifier import Input
from src.backend.PluginManager.ActionCore import ActionCore
from src.backend.PluginManager.EventAssigner import EventAssigner

from ...xplane import (
	apply_index,
	apply_step,
	coerce_number,
	format_value,
	parse_dataref_path,
)

_DIR_SIGN = {
	"right": 1,
	"up": 1,
	"left": -1,
	"down": -1,
}


class Rotary(ActionCore):
	def __init__(self, *args, **kwargs):
		super().__init__(*args, **kwargs)
		self.has_configuration = True
		self._last_value: float | None = None
		self._hold_id: int | None = None

		self.add_event_assigner(
			EventAssigner(
				id="xplane_rotary_down",
				ui_label="Press",
				default_events=[Input.Key.Events.DOWN, Input.Dial.Events.DOWN],
				callback=self.on_press,
			)
		)
		self.add_event_assigner(
			EventAssigner(
				id="xplane_rotary_up",
				ui_label="Release",
				default_events=[Input.Key.Events.UP, Input.Dial.Events.UP],
				callback=self.on_release,
			)
		)

	def on_ready(self) -> None:
		self.set_media(media_path=self.get_asset_path("info.png"), size=0.75)
		s = self.get_settings()
		self.set_top_label((s.get("label") or "").strip())
		self._refresh_label()

	def on_tick(self) -> None:
		self._refresh_label()

	def get_config_rows(self):
		EntryRow(
			action_core=self,
			var_name="command_path",
			default_value="",
			title="Command Path",
			on_change=lambda *_: self.on_ready(),
		)
		ComboRow(
			action_core=self,
			var_name="direction",
			default_value="right",
			title="Direction",
			items=["right", "left", "up", "down"],
		)
		EntryRow(
			action_core=self,
			var_name="dataref_path",
			default_value="sim/cockpit2/engine/actuators/ignition_key",
			title="DataRef Path",
			on_change=lambda *_: self.on_ready(),
		)
		SpinRow(
			action_core=self,
			var_name="delta",
			default_value=0.0,
			title="Delta (0 = use Command Path)",
			min=0.0,
			max=1_000_000.0,
			step=0.01,
			digits=4,
		)
		EntryRow(
			action_core=self,
			var_name="min_value",
			default_value="",
			title="Min Value",
		)
		EntryRow(
			action_core=self,
			var_name="max_value",
			default_value="",
			title="Max Value",
		)
		SwitchRow(
			action_core=self,
			var_name="cycle",
			default_value=False,
			title="Cycle at Min/Max",
		)
		SwitchRow(
			action_core=self,
			var_name="hide_endstop_alert",
			default_value=False,
			title="Hide endstop alert",
		)
		EntryRow(
			action_core=self,
			var_name="hold_command",
			default_value="",
			title="Hold Command (at last position)",
		)
		SwitchRow(
			action_core=self,
			var_name="hold_on_last",
			default_value=False,
			title="Hold when already at last position",
		)
		EntryRow(
			action_core=self,
			var_name="label",
			default_value="",
			title="Label",
			on_change=lambda *_: self.on_ready(),
		)
		EntryRow(
			action_core=self,
			var_name="format",
			default_value="%s",
			title="Format",
		)
		EntryRow(
			action_core=self,
			var_name="unit",
			default_value="",
			title="Unit",
		)
		return self.get_generative_ui_widgets()

	def on_press(self, _data=None) -> None:
		s = self.get_settings()
		delta = float(s.get("delta") or 0)
		dataref = (s.get("dataref_path") or "").strip()
		command = (s.get("command_path") or "").strip()
		hold_cmd = (s.get("hold_command") or "").strip()
		can_step = delta > 0 and bool(dataref)

		if not can_step and not command and not hold_cmd:
			self.show_error(duration=1)
			return

		if can_step and bool(s.get("hold_on_last", False)) and hold_cmd:
			if self._at_last_position(delta):
				self._begin_hold(hold_cmd)
				return

		if can_step:
			self._step_dataref(delta)
			return

		if command:
			self._fire_command(command)
		else:
			self.show_error(duration=1)

	def on_release(self, _data=None) -> None:
		if self._hold_id is None:
			return
		try:
			self.plugin_base.xplane.end_command(self._hold_id)
			log.info(f"rotary hold end id={self._hold_id}")
		except Exception as err:
			log.error(f"rotary hold end failed: {err}")
			self.show_error(duration=2)
		finally:
			self._hold_id = None

	def _sign(self) -> int:
		direction = (self.get_settings().get("direction") or "right").strip().lower()
		return _DIR_SIGN.get(direction, 1)

	def _opt_float(self, key: str) -> float | None:
		raw = self.get_settings().get(key, "")
		if raw in (None, ""):
			return None
		try:
			return float(raw)
		except (TypeError, ValueError):
			return None

	def _at_last_position(self, delta: float) -> bool:
		"""True when the next step would be blocked (endstop) without cycle."""
		s = self.get_settings()
		if bool(s.get("cycle", False)):
			return False
		min_v = self._opt_float("min_value")
		max_v = self._opt_float("max_value")
		path = (s.get("dataref_path") or "").strip()
		if not path:
			return False
		try:
			xplane = self.plugin_base.xplane
			base_path, index = parse_dataref_path(path)
			dr_id = xplane.get_dataref_id(base_path)
			current_raw = (
				self._last_value
				if self._last_value is not None
				else apply_index(xplane.read_dataref(dr_id), index)
			)
			current = coerce_number(current_raw) or 0.0
			_, blocked = apply_step(
				current,
				self._sign() * delta,
				min_v=min_v,
				max_v=max_v,
				cycle=False,
			)
			return blocked
		except Exception:
			return False

	def _begin_hold(self, path: str) -> None:
		try:
			xplane = self.plugin_base.xplane
			cmd_id = xplane.get_command_id(path)
			xplane.begin_command(cmd_id)
			self._hold_id = cmd_id
			log.info(f"rotary hold begin: {path} (id={cmd_id})")
		except Exception as err:
			log.error(f"rotary hold begin failed: {path}: {err}")
			self.show_error(duration=2)

	def _fire_command(self, path: str) -> None:
		try:
			xplane = self.plugin_base.xplane
			cmd_id = xplane.get_command_id(path)
			xplane.activate_command(cmd_id)
			log.info(f"rotary command: {path}")
		except Exception as err:
			log.error(f"rotary command failed: {path}: {err}")
			self.show_error(duration=2)

	def _step_dataref(self, delta: float) -> None:
		s = self.get_settings()
		path = (s.get("dataref_path") or "").strip()
		min_v = self._opt_float("min_value")
		max_v = self._opt_float("max_value")
		cycle = bool(s.get("cycle", False))
		hide_endstop = bool(s.get("hide_endstop_alert", False))
		step = self._sign() * delta

		try:
			xplane = self.plugin_base.xplane
			base_path, index = parse_dataref_path(path)
			dr_id = xplane.get_dataref_id(base_path)
			current_raw = (
				self._last_value
				if self._last_value is not None
				else apply_index(xplane.read_dataref(dr_id), index)
			)
			current = coerce_number(current_raw) or 0.0
			target, blocked = apply_step(
				current, step, min_v=min_v, max_v=max_v, cycle=cycle
			)
			if blocked:
				log.info(f"rotary endstop {path} value={current}")
				if not hide_endstop:
					self.show_error(duration=1)
				return
			xplane.write_dataref(dr_id, target, index)
			self._last_value = target
			self._paint_value(target)
			log.info(f"rotary step {path}: {current} → {target}")
		except Exception as err:
			log.error(f"rotary step failed: {err}")
			self.show_error(duration=2)

	def _refresh_label(self) -> None:
		s = self.get_settings()
		path = (s.get("dataref_path") or "").strip()
		if not path:
			self.set_center_label("")
			return
		try:
			xplane = self.plugin_base.xplane
			base_path, index = parse_dataref_path(path)
			dr_id = xplane.get_dataref_id(base_path)
			raw = apply_index(xplane.read_dataref(dr_id), index)
			num = coerce_number(raw)
			if num is not None:
				self._last_value = num
			self._paint_value(raw)
		except Exception:
			offline = not self.plugin_base.xplane.ping()
			self.set_center_label("OFFLINE" if offline else "?")

	def _paint_value(self, raw) -> None:
		s = self.get_settings()
		text = format_value(
			raw,
			fmt=(s.get("format") or "%s").strip() or "%s",
			unit=(s.get("unit") or "").strip(),
		)
		self.set_center_label(text)
